# Notebook 1: Getting Started with Python

Welcome to your first Python notebook! In this session, you'll learn the fundamental building blocks of Python programming that you'll need for data science.

**Learning Objectives:**
- Understand variables and how to assign values
- Work with different data types (numbers, strings, booleans)
- Perform basic arithmetic operations
- Use print statements for output
- Write comments to document your code

## What is a Jupyter Notebook?

A Jupyter notebook combines text (like this paragraph) with executable code. You can run code by clicking the "Run" button or pressing Shift+Enter. Let's start with a simple example:

In [6]:
# This is a comment - Python ignores lines that start with #
print("Hello, Python!")

Hello, Python!


## Variables and Assignment

Variables are like containers that store data. In Python, you create a variable by assigning a value to it using the `=` operator.

In [ ]:
# Creating variables
name = "Alice"
age = 25
height = 5.6

# Display the values using f-strings (modern Python)
print(f"Name: {name}")
print(f"Age: {age}")
print(f"Height: {height}")

Name: Alice
Age: 25
Height: 5.6


## Data Types

Python has several built-in data types. Let's explore the most important ones for data science:

### Numbers (Integers and Floats)

In [ ]:
# Integers (whole numbers)
count = 42
temperature = -5

# Floats (decimal numbers)
price = 19.99
pi = 3.14159

print(f"Count: {count}, Type: {type(count)}")
print(f"Price: {price}, Type: {type(price)}")

Count: 42 Type: <class 'int'>
Price: 19.99 Type: <class 'float'>


### Strings (Text)

In [ ]:
# Strings can use single or double quotes
message = "Welcome to Python!"
city = 'New York'

print(message)
print(city)
print(f"Type of message: {type(message)}")

Welcome to Python!
New York
Type of message: <class 'str'>


### Booleans (True/False)

In [ ]:
# Boolean values
is_student = True
is_working = False

print(f"Is student: {is_student}")
print(f"Is working: {is_working}")
print(f"Type: {type(is_student)}")

Is student: True
Is working: False
Type: <class 'bool'>


## Arithmetic Operations

Python supports all basic mathematical operations:

In [ ]:
# Basic arithmetic
a = 10
b = 3

print(f"Addition: {a + b}")
print(f"Subtraction: {a - b}")
print(f"Multiplication: {a * b}")
print(f"Division: {a / b}")
print(f"Integer Division: {a // b}")  # Rounds down to nearest integer
print(f"Remainder (Modulo): {a % b}")
print(f"Exponentiation: {a ** b}")  # a to the power of b

Addition: 13
Subtraction: 7
Multiplication: 30
Division: 3.3333333333333335
Integer Division: 3
Remainder (Modulo): 1
Exponentiation: 1000


## String Operations

Strings have many useful operations that you'll use in data science:

In [ ]:
first_name = "John"
last_name = "Doe"

# String concatenation (joining)
full_name = first_name + " " + last_name
print(f"Full name: {full_name}")

# String length
print(f"Length of full name: {len(full_name)}")

# String methods
print(f"Uppercase: {full_name.upper()}")
print(f"Lowercase: {full_name.lower()}")

Full name: John Doe
Length of full name: 8
Uppercase: JOHN DOE
Lowercase: john doe


## String Formatting with f-strings

String formatting is crucial for displaying results in data science. We'll use f-strings (formatted string literals), which are the modern Python way to format strings. F-strings are faster, more readable, and easier to use than older methods.

In [ ]:
name = "Alice"
score = 85.7

# Using f-strings (modern Python - recommended approach)
print(f"Student {name} scored {score:.1f}%")

# Additional f-string formatting examples
print(f"Student {name} has a score of {score}")
print(f"Rounded score: {score:.0f}")
print(f"Score with 2 decimals: {score:.2f}")

Student Alice scored 85.7%
Student Alice has a score of 85.7
Rounded score: 86
Score with 2 decimals: 85.70


### 🎯 Understanding F-String Formatting in Detail

Let's break down exactly how the f-string formatting works in the code above:

#### The Variables
- `name = "Alice"` → A **string** containing the text "Alice"
- `score = 85.7` → A **float** (decimal number) with the value 85.7

#### F-String Breakdown

**1. Basic F-String with Number Formatting:**
```python
print(f"Student {name} scored {score:.1f}%")
```
- `f"..."` → Creates an **f-string** (formatted string literal)
- `{name}` → Inserts the value of the `name` variable directly
- `{score:.1f}` → Inserts the `score` variable with special formatting:
  - `:` → Starts the format specification
  - `.1` → Show exactly 1 decimal place
  - `f` → Format as a floating-point number
- `%` → A literal percent sign (not part of formatting)

**2. Simple Variable Insertion:**
```python
print(f"Student {name} has a score of {score}")
```
No special formatting - just inserts the variables as-is.

**3. Rounded to Whole Number:**
```python
print(f"Rounded score: {score:.0f}")
```
`{score:.0f}` → Show 0 decimal places (rounds 85.7 to 86)

**4. Two Decimal Places:**
```python
print(f"Score with 2 decimals: {score:.2f}")
```
`{score:.2f}` → Show exactly 2 decimal places (adds trailing zero: 85.70)

#### 🔧 Common F-String Format Patterns for Data Science

```python
# Number formatting examples
value = 1234.5678
percentage = 0.847
large_number = 1500000

# Different decimal places
print(f"{value:.2f}")           # 1234.57 (2 decimals)
print(f"{value:.0f}")           # 1235 (rounded to whole)

# Thousands separator
print(f"{large_number:,}")      # 1,500,000 (with commas)
print(f"{large_number:,.2f}")   # 1,500,000.00 (commas + decimals)

# Percentage formatting
print(f"{percentage:.1%}")      # 84.7% (auto converts to %)
print(f"{percentage:.2%}")      # 84.70%

# Padding and alignment
print(f"{value:10.2f}")         # "   1234.57" (right-aligned in 10 chars)
print(f"{value:<10.2f}")        # "1234.57   " (left-aligned)
```

#### 💡 Why F-Strings Are Essential for Data Science

1. **Professional Reports:** Create clean, formatted output for stakeholders
2. **Readability:** Variables are clearly visible in the string
3. **Performance:** Faster than older formatting methods
4. **Consistency:** Ensures numbers display with appropriate precision
5. **Debugging:** Easy to see what variables are being used

**You'll use f-strings constantly** in data science for displaying analysis results, creating file names, formatting reports, and presenting findings!

## Working with Variables

Variables can be updated and used in calculations:

In [ ]:
# Starting values
balance = 1000
interest_rate = 0.05

print(f"Initial balance: {balance}")

# Update the balance
balance = balance + (balance * interest_rate)
print(f"Balance after interest: {balance}")

# Shorthand for updating variables
balance += 100  # Same as: balance = balance + 100
print(f"Balance after deposit: {balance}")

Initial balance: 1000
Balance after interest: 1050.0
Balance after deposit: 1150.0


## Data Analysis Example: Investment Portfolio

Let's apply what we've learned to a realistic data science scenario - analyzing investment returns:

In [ ]:
# Real-world scenario: Analyzing investment portfolio performance
# Data for three different investments
stock_a_price = 150.75
stock_a_shares = 50
stock_a_dividend = 2.30

stock_b_price = 89.20
stock_b_shares = 100
stock_b_dividend = 1.85

bond_investment = 5000
bond_interest_rate = 0.035

# Calculate portfolio values
stock_a_value = stock_a_price * stock_a_shares
stock_b_value = stock_b_price * stock_b_shares
total_stock_value = stock_a_value + stock_b_value

# Calculate annual income from dividends and bonds
stock_a_annual_income = stock_a_dividend * stock_a_shares
stock_b_annual_income = stock_b_dividend * stock_b_shares
bond_annual_income = bond_investment * bond_interest_rate
total_annual_income = stock_a_annual_income + stock_b_annual_income + bond_annual_income

# Generate investment report (like a data scientist would)
print("=== INVESTMENT PORTFOLIO ANALYSIS ===")
print(f"Stock A Portfolio Value: {stock_a_value}")
print(f"Stock A Portfolio Value: ${stock_a_value:,.2f}")
print(f"Stock B Portfolio Value: ${stock_b_value:,.2f}")
print(f"Bond Investment: ${bond_investment:,.2f}")
print(f"Total Portfolio Value: ${total_stock_value + bond_investment:,.2f}")
print()
print("=== ANNUAL INCOME PROJECTION ===")
print(f"Stock A Dividends: ${stock_a_annual_income:.2f}")
print(f"Stock B Dividends: ${stock_b_annual_income:.2f}")
print(f"Bond Interest: ${bond_annual_income:.2f}")
print(f"Total Annual Income: ${total_annual_income:.2f}")
print(f"Income Yield: {(total_annual_income / (total_stock_value + bond_investment)) * 100:.2f}%")

=== INVESTMENT PORTFOLIO ANALYSIS ===
Stock A Portfolio Value: 7537.5
Stock A Portfolio Value: $7,537.50
Stock B Portfolio Value: $8,920.00
Bond Investment: $5,000.00
Total Portfolio Value: $21,457.50

=== ANNUAL INCOME PROJECTION ===
Stock A Dividends: $115.00
Stock B Dividends: $185.00
Bond Interest: $175.00
Total Annual Income: $475.00
Income Yield: 2.21%


## Practice Exercises

Try these exercises to practice what you've learned:

### Exercise 1: Personal Information
Create variables for your name, age, and favorite number. Print them in a formatted sentence.

In [ ]:
# Your code here
my_name = "Your Name"
my_age = 0
favorite_number = 0

# Print a sentence using these variables


### Exercise 2: Simple Calculator
Create two number variables and perform all arithmetic operations on them.

In [ ]:
# Your code here
num1 = 15
num2 = 4

# Perform and print all arithmetic operations


### Exercise 3: Data Analysis Scenario
You're analyzing test scores. Calculate the average of three scores and determine if it's a passing grade (>= 70).

In [ ]:
# Three test scores
score1 = 78
score2 = 85
score3 = 92

# Calculate average
average = (score1 + score2 + score3) / 3

# Determine if passing (we'll learn better ways to do this in the next notebook)
is_passing = average >= 70

print(f"Average score: {average:.1f}")
print(f"Is passing: {is_passing}")

Average score: 85.0
Is passing: True


## Key Takeaways

1. **Variables** store data and can be updated
2. **Data types** include integers, floats, strings, and booleans
3. **Arithmetic operations** work as expected with numbers
4. **String formatting** is essential for displaying results
5. **Comments** (using #) help document your code

These fundamentals will be used throughout your data science journey. In the next notebook, we'll learn about control structures that let you make decisions and repeat operations in your code.

---

## 📝 Mini-Challenge: Practice What You've Learned

Test your understanding with these hands-on exercises:

### Challenge 1: Personal Data Calculator
Create variables for:
- Your name (string)
- Your age (integer) 
- Your height in meters (float)
- Whether you like Python (boolean)

Then calculate and print:
- Your age in months
- Your height in centimeters
- A message combining your name and whether you like Python

In [ ]:
# Challenge 1: Your solution here
# Create your variables
my_name = "Your Name"  # Replace with your actual name
my_age = 25           # Replace with your actual age
my_height_m = 1.75    # Replace with your actual height in meters
likes_python = True   # Do you like Python?

# Calculate values
age_in_months = my_age * 12
height_in_cm = my_height_m * 100

# Print results
print(f"Name: {my_name}")
print(f"Age in months: {age_in_months}")
print(f"Height in cm: {height_in_cm}")
print(f"{my_name} likes Python: {likes_python}")

Name: Your Name
Age in months: 300
Height in cm: 175.0
Your Name likes Python: True


### Challenge 2: Data Type Explorer
Create variables of different types and use `type()` to check their types. Try some edge cases:
- What happens when you divide two integers?
- What's the type of `3.0`?
- Try creating a string with quotes inside it

In [ ]:
# Challenge 2: Your solution here
# Experiment with different data types
number1 = 10
number2 = 3
result = number1 / number2
print(f"10 / 3 = {result}, type: {type(result)}")

float_number = 3.0
print(f"3.0 type: {type(float_number)}")

# String with quotes inside
quote_string = "She said, 'Hello, Python!'"
print(f"String with quotes: {quote_string}")

# Try your own experiments below:

10 / 3 = 3.3333333333333335, type: <class 'float'>
3.0 type: <class 'float'>
String with quotes: She said, 'Hello, Python!'


---

## ✅ Self-Assessment Checklist

Before moving to the next notebook, make sure you can:

- [ ] Create variables and assign different types of values
- [ ] Understand the difference between integers, floats, strings, and booleans
- [ ] Use arithmetic operators (+, -, *, /, //, %, **)
- [ ] Write and understand comments in code
- [ ] Use print() statements to display output
- [ ] Use f-strings for formatted output
- [ ] Check the type of a variable using type()

**Pro Tip:** If you're unsure about any item, review the relevant section above or ask for help!

---

## 🚀 What's Next?

In the next notebook, you'll learn about:
- **Control Structures**: Making decisions with if/else statements
- **Loops**: Repeating actions efficiently
- **User Input**: Making your programs interactive

These are essential tools for writing more complex data science programs!